astropy/cosmology/connect.py

```diff
@@ -237,8 +322,56 @@ class CosmologyToFormat(io_registry.UnifiedReadWrite):
         Keyword arguments passed through to data writer.
     """
 
-    def __init__(self, instance, cls):
+    def __init__(self, instance: Cosmology, cls: type[Cosmology]) -> None:
         super().__init__(instance, cls, "write", registry=convert_registry)
 
-    def __call__(self, format, *args, **kwargs):
+    # ===============================================================
+    # __call__ overloads
+
+    # To a Cosmology
+    @overload
+    def __call__(
+        self, format: Literal["astropy.cosmology"], *args: Any, **kwargs: Any
+    ) -> Cosmology:
+        ...
+
+    # To a Model
+    @overload
+    def __call__(
+        self, format: Literal["astropy.model"], *args: Any, **kwargs: Any
+    ) -> _CosmologyModel:
+        ...
+
+    # To a Row
+    @overload
+    def __call__(
+        self, format: Literal["astropy.row"], *args: Any, **kwargs: Any
+    ) -> Row:
+        ...
+
+    # To a Table
+    @overload
+    def __call__(
+        self, format: Literal["astropy.table"], *args: Any, **kwargs: Any
+    ) -> Table:
+        ...
+
+    # To a Mapping
+    @overload
+    def __call__(
+        self, format: Literal["mapping"], *args: Any, **kwargs: Any
+    ) -> Mapping[str, Any]:
```

Comment:
[`Mapping`](https://docs.python.org/3/library/collections.abc.html#collections.abc.Mapping) is an abstract class, so the method must be returning an instance of some more specific class. Why aren't we using that as the annotation?

Reply:
The base type was the best that could be done.

Reply:
There will be cases where it is possible to spend a lot of effort to make the annotations more specific, but the benefit of doing so is very small. It could well be that this is one of those cases. But _if_ this can be easily solved with more overloads then we should consider that.

Reply:
Hard to tell if this is exactly correct. ``mypy astropy/cosmology/connect.py --follow-imports=skip --strict --show-error-codes`` is no longer complaining, but that's not a great guarantee sans actual mypy config project-wide.

Reply:
The overloads look good enough to me.